main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -0,0 +1,852 @@
+   * which needs to be sign extended to represent that value as an Int32 (int).
+   *
+   * Example:
+   * Consider the following (signed) right shift, where x represents an unknown value (0 or 1):
+   *   x0000000 >> 4 = xxxxx000
+   * These Int8s (bytes) are represented as Int32s (ints), so the x is sign extended:
+   *   xxxxxxxx xxxxxxxx xxxxxxxx x0000000 >> 4 = xxxxxxxx xxxxxxxx xxxxxxxx xxxxx000
+   *
+   * We don't need to truncate, because it is impossible for random data to be in the higher-order bits. Either those
+   * bits are all 0, or they are 1 (because of sign extension).
+   *
+   * Note: the right-hand operand of a shift (i.e. the shift amount) *must* be Int32.
+   */
+  private def compileBitwiseExpr(currentClassType: JvmType.Reference,
+                                 visitor: MethodVisitor,
+                                 jumpLabels: Map[Symbol.LabelSym, Label],
+                                 entryPoint: Label)
+                                (o: BitwiseOperator, e1: Expression, e2: Expression)(implicit root: Root, flix: Flix): Unit = {
+    compileExpression(currentClassType, e1, jumpLabels, entryPoint, visitor)
+    compileExpression(currentClassType, e2, jumpLabels, entryPoint, visitor)
+    val (intOp, longOp, bigintOp) = o match {
+      case BinaryOperator.BitwiseAnd => (IAND, LAND, "and")
+      case BinaryOperator.BitwiseOr => (IOR, LOR, "or")
+      case BinaryOperator.BitwiseXor => (IXOR, LXOR, "xor")
+      case BinaryOperator.BitwiseLeftShift => (ISHL, LSHL, "shiftLeft")
+      case BinaryOperator.BitwiseRightShift => (ISHR, LSHR, "shiftRight")
+    }
+    e1.tpe match {
+      case Type.Int8 =>
+        visitor.visitInsn(intOp)
+        if (intOp == ISHL) visitor.visitInsn(I2B)
+      case Type.Int16 =>
+        visitor.visitInsn(intOp)
+        if (intOp == ISHL) visitor.visitInsn(I2S)
+      case Type.Int32 => visitor.visitInsn(intOp)
+      case Type.Int64 => visitor.visitInsn(longOp)
+      case Type.BigInt =>
+        visitor.visitMethodInsn(INVOKEVIRTUAL, JvmName.BigInteger.toInternalName,
+          bigintOp, AsmOps.getMethodDescriptor(List(JvmType.BigInteger), JvmType.BigInteger), false)
+      case _ => throw InternalCompilerException(s"Can't apply $o to type ${e1.tpe}.")
+    }
+  }
+
+  /*
+   * Adding the source of the line for debugging
+   */
+  private def addSourceLine(visitor: MethodVisitor, loc: SourceLocation): Unit = {
+    val label = new Label()
+    visitor.visitLabel(label)
+    visitor.visitLineNumber(loc.beginLine, label)
+  }
+
+  /*
+   * `tpe` is type of value on top of the stack. If the value is not primitive, then we cast it to it's specific type,
+   * if the value is a primitive then since there is no boxing, then no casting is necessary.
+   */
+  private def castIfNotPrim(tpe: Type, visitor: MethodVisitor)(implicit root: Root, flix: Flix): Unit = tpe match {
```

Comment:
I think either this should be rewritten in terms of JvmType or moved into JvmOps as a primitive.